vulkan/vulkan_handle.h

```diff
@@ -0,0 +1,71 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_VULKAN_VULKAN_HANDLE_H_
+#define FLUTTER_VULKAN_VULKAN_HANDLE_H_
+
+#include <functional>
+
+#include "lib/ftl/macros.h"
+#include "vulkan_interface.h"
+#include "lib/ftl/logging.h"
+
+namespace vulkan {
+
+template <class T>
+class VulkanHandle {
+ public:
+  using Handle = T;
+  using Disposer = std::function<void(Handle)>;
+
+  VulkanHandle() : handle_(VK_NULL_HANDLE) {}
+
+  VulkanHandle(Handle handle, Disposer disposer)
+      : handle_(handle), disposer_(disposer) {}
+
+  VulkanHandle(VulkanHandle&& other)
+      : handle_(other.handle_), disposer_(other.disposer_) {
+    other.handle_ = VK_NULL_HANDLE;
+    other.disposer_ = nullptr;
+  }
+
+  ~VulkanHandle() { DisposeIfNecessary(); }
+
+  VulkanHandle& operator=(VulkanHandle&& other) {
+    if (handle_ != other.handle_) {
+      DisposeIfNecessary();
+    }
+
+    handle_ = other.handle_;
+    disposer_ = other.disposer_;
+
+    other.handle_ = VK_NULL_HANDLE;
+    other.disposer_ = nullptr;
```

Comment:
This will still mess up under self-assignment.

I'd recommend swapping the handle and disposer of `other` into temporaries.  Then calling DisposeIfNecessary, then assigning the temporaries into the member variables.